datafusion-examples/examples/simple_udf.rs

```diff
@@ -61,17 +62,24 @@ async fn main() -> Result<()> {
     let ctx = create_context()?;
 
     // First, declare the actual implementation of the calculation
-    let pow = |args: &[ArrayRef]| {
+    let pow = Arc::new(|args: &[ColumnarValue]| {
         // in DataFusion, all `args` and output are dynamically-typed arrays, which means that we need to:
         // 1. cast the values to the type we want
         // 2. perform the computation for every element in the array (using a loop or SIMD) and construct the result
 
         // this is guaranteed by DataFusion based on the function's signature.
         assert_eq!(args.len(), 2);
 
+        let ColumnarValue::Array(arg0) = &args[0] else {
```

Comment:
I think the more correct pattern is to create an array with [`into_array`](https://docs.rs/datafusion/latest/datafusion/physical_plan/enum.ColumnarValue.html#method.into_array) 

```suggestion
        let arg0 = args[0].into_array(num_rows)
```

Though now I wrote that it is not super clear to me where the num_rows should come from 🤔 

Also, maybe this is more of the fix for https://github.com/apache/arrow-datafusion/issues/8866 -- that for a volatile function, it shouldn't be passed a scalar but unstead the array should be expanded prior to invocation

Reply:
In what situations would 'arg' be a Scalar? If we can provide an example to reproduce it, our case/test can cover more condition?

Reply:
I think something like `pow(1.0, 2.0)` (pass in literal values)